### PR TITLE
Refactor Package Repository

### LINKED ISSUE #73: Workspace should show a suitable error message when local harnesses.json source is not found
When you specify a local `harnesses.json` source in your `workspace.yml`, and that file is non-existent, workspace gives  the following output:

```
PHP Warning:  file_get_contents(./tools/harnesses.json): failed to open stream: No such file or directory in phar:///usr/local/bin/ws/src/Types/Harness/Repository/Repository.php on l
ine 78

Warning: file_get_contents(./tools/harnesses.json): failed to open stream: No such file or directory in phar:///usr/local/bin/ws/src/Types/Harness/Repository/Repository.php on line 7
8
PHP Warning:  array_merge(): Expected parameter 2 to be an array, null given in phar:///usr/local/bin/ws/src/Types/Harness/Repository/Repository.php on line 78

Warning: array_merge(): Expected parameter 2 to be an array, null given in phar:///usr/local/bin/ws/src/Types/Harness/Repository/Repository.php on line 78
PHP Notice:  Trying to access array offset on value of type null in phar:///usr/local/bin/ws/src/Types/Harness/Repository/Repository.php on line 91

Notice: Trying to access array offset on value of type null in phar:///usr/local/bin/ws/src/Types/Harness/Repository/Repository.php on line 91
PHP Warning:  array_keys() expects parameter 1 to be array, null given in phar:///usr/local/bin/ws/src/Types/Harness/Repository/Repository.php on line 91

Warning: array_keys() expects parameter 1 to be array, null given in phar:///usr/local/bin/ws/src/Types/Harness/Repository/Repository.php on line 91
PHP Warning:  Invalid argument supplied for foreach() in phar:///usr/local/bin/ws/src/Types/Harness/Repository/Repository.php on line 99

Warning: Invalid argument supplied for foreach() in phar:///usr/local/bin/ws/src/Types/Harness/Repository/Repository.php on line 99
PHP Fatal error:  Uncaught Exception: Could not resolve 'inviqa/go:v0.7.0' to a harness package' in phar:///usr/local/bin/ws/src/Types/Harness/Repository/Repository.php:125
```

It would be nicer to have workspace validate the file exists and bail out with a suitable error if it is no there.

### PR DESCRIPTION
**Pull request by dantleech**
Fixes #73 

This PR:

- Adds a test for the package repository
- Handles runtime errors in the package repository
- Will introduce a `FileLoader` which will safely handle file loading (from URLs)

### CODE REVIEW

File: src/Types/Harness/Repository/PackageRepository.php [hunk 1]
```diff
@@ -19,17 +21,23 @@ class PackageRepository implements Repository
...
 
         foreach (['name', 'version', 'dist'] as $name) {
             $this->properties[$name] = new ReflectionProperty(Package::class, $name);
             $this->properties[$name]->setAccessible(true);
         }
+        $this->fileLoader = $fileLoader;
     }
 
-    public function addPackage($name, $version, $dist)
+    public function addPackage(string $name, string $version, array $dist)
```
Comment (dantleech) [inline src/Types/Harness/Repository/PackageRepository.php]: void (although is it part of hte interface?)

File: src/Types/Harness/Repository/PackageRepository.php [hunk 2]
```diff
@@ -45,6 +53,13 @@ public function get(string $package): Package
 
         preg_match(self::HARNESS_PACKAGE_PATTERN, $package, $match);
 
+        if (!isset($match['harness'])) {
```
Comment (dantleech) [inline src/Types/Harness/Repository/PackageRepository.php]: could check the result of preg_match instead?

File: src/Types/Harness/Repository/PackageRepository.php [hunk 3]
```diff
@@ -86,42 +101,50 @@ private function resolvePackageVersion(string $name, string $version): string
...
+        if (!isset($this->packages[$name])) {
+            throw new RuntimeException(sprintf(
+                'Package "%s" is not registered, registered packages "%s"',
+                $name, implode('", "', array_keys($this->packages))
+            ));
+        }
 
-            $collection = array_keys($this->packages[$name]);
+        if (!preg_match(self::HARNESS_VERSION_PATTERN, $version, $match)) {
+            throw new RuntimeException(sprintf('Invalid version string "%s"', $version));
```
Comment (dantleech) [inline src/Types/Harness/Repository/PackageRepository.php]: not sure why we need this twice

File: src/Types/Harness/Repository/PackageRepository.php [hunk 4]
```diff
@@ -86,42 +101,50 @@ private function resolvePackageVersion(string $name, string $version): string
...
+                $candidate = $availVersion;
             }
         }
 
-        throw new Exception("Could not resolve '{$name}:{$version}' to a harness package'");
+        if ($candidate === null) {
+            throw new Exception("Could not resolve '{$name}:{$version}' to a harness package'");
+        }
+
+        return $availVersion;
```
Comment (dantleech) [inline src/Types/Harness/Repository/PackageRepository.php]: removed the "big if" statement above in favor of a guard condition

File: src/Types/Harness/Repository/PackageRepository.php [hunk 5]
```diff
@@ -56,6 +61,18 @@ public function get(string $package): Package
         ]);
     }
 
+    public function addPackage(string $name, string $version, array $dist): void
+    {
+        $this->parseVersionString($version);
+        $this->packages[$name][$version] = $dist;
+    }
+
+    public function addSource($url): void
```
Comment (dantleech) [inline src/Types/Harness/Repository/PackageRepository.php]: can add type hint

File: src/Types/Harness/Repository/PackageRepository.php [hunk 6]
```diff
@@ -67,61 +84,86 @@ private function hydrate(array $values): Package
...
+                    $source['url']
+                ), 0, $error);
+            }
             $this->sources[$k]['imported'] = true;
         }
     }
 
     private function resolvePackageVersion(string $name, string $version): string
     {
+        [$major, $minor, $patch] = $this->parseVersionString($version);
```
Comment (dantleech) [inline src/Types/Harness/Repository/PackageRepository.php]: `mustParseVersionString`

File: src/Types/Harness/Repository/PackageRepository.php [hunk 7]
```diff
@@ -67,61 +84,86 @@ private function hydrate(array $values): Package
...
-        throw new Exception("Could not resolve '{$name}:{$version}' to a harness package'");
+        if ($candidate === null) {
+            throw new Exception(sprintf('Could not resolve "%s:%s" to a harness package', $name, $version));
+        }
+
+        return $candidate;
+    }
+
+    /**
+     * @return array{int,int,int}
```
Comment (dantleech) [inline src/Types/Harness/Repository/PackageRepository.php]: is `int|'x'`

**Review by jameshalsall — DISMISSED** [review]
LGTM, nice work 👍🏻

**Review by jameshalsall — APPROVED** [review]

**Pull request merged**